src/zarr/codecs/mixins.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import Awaitable, Callable, Generic, Iterable, TypeVar
+
+import numpy as np
+
+from zarr.abc.codec import (
+    ArrayArrayCodec,
+    ArrayBytesCodec,
+    ArrayBytesCodecPartialDecodeMixin,
+    ArrayBytesCodecPartialEncodeMixin,
+    ByteGetter,
+    ByteSetter,
+    BytesBytesCodec,
+)
+from zarr.common import ArraySpec, BytesLike, SliceSelection, concurrent_map
+from zarr.config import config
+
+
+CodecInput = TypeVar("CodecInput", bound=np.ndarray | BytesLike)
+CodecOutput = TypeVar("CodecOutput", bound=np.ndarray | BytesLike)
+
+
+async def batching_helper(
```

Comment:
This doesn't need to be addressed here, so feel free to ignore for now

Reply:
How would you feel about this?
```python
class _Codec(Generic[CodecInput, CodecOutput], Metadata):
    ...

    async def _decode_single(self, chunk_data: CodecOutput, chunk_spec: ArraySpec) -> CodecInput:
        raise NotImplementedError

    async def decode(
        self, chunk_data_and_specs: Iterable[tuple[CodecOutput | None, ArraySpec]]
    ) -> Iterable[CodecInput | None]:
        return await batching_helper(self.decode_single, chunk_data_and_specs)

    # same for encode
    ...
```
Batch-aware codecs would then only override the `decode` method and ignore the `_decode_single` method. That would be fine because `_decode_single` is a protected method not intended to be part of the public interface. Single-chunk codecs could override `_decode_single` and won't have to care about the batching. We could drop the batch mixins, then.

Reply:
sounds good!